In [301]:
import os
print("Current directory:", os.getcwd())

Current directory: d:\WORKSPACE\DATA-SCIENCE\ML\PROJECTS\spotify-ai-recommender\notebooks


In [302]:
import pandas as pd


In [303]:
df1 = pd.read_csv("../data/raw/spotify_data clean.csv")
#df2 = pd.read_csv("../data/raw/track_data_final.csv")

In [304]:
df1.head()

track_id                                       track_name  \
0  3EJS5LyekDim1Tf5rBFmZl                    Trippy Mane (ft. Project Pat)   
1  1oQW6G2ZiwMuHqlPpP27DB                                             OMG!   
2  7mdkjzoIYlf1rx9EtBpGmU                                      Hard 2 Find   
3  67rW0Zl7oB3qEpD5YWWE5w  Still Get Like That (ft. Project Pat & Starrah)   
4  15xptTfRBrjsppW0INUZjf                            ride me like a harley   

   track_number  track_popularity  explicit artist_name  artist_popularity  \
0             4                 0      True       Diplo                 77   
1             1                 0      True    Yelawolf                 64   
2             1                 4      True   Riff Raff                 48   
3             8                30      True       Diplo                 77   
4             2                 0      True     Rumelis                 48   

   artist_followers                      artist_genres  \
0           2812821                         moombahton   
1           2363438  country hip hop, southern hip hop   
2            193302                                NaN   
3           2813710                         moombahton   
4              8682                           dark r&b   

                 album_id                           album_name  \
0  5QRFnGnBeMGePBKF2xTz5z                   d00mscrvll, Vol. 1   
1  4SUmmwnv0xTjRcLdjczGg2                                 OMG!   
2  3E3zEAL8gUYWaLYB9L7gbp                          Hard 2 Find   
3  5QRFnGnBeMGePBKF2xTz5z                   d00mscrvll, Vol. 1   
4  06FDIpSHYmZAZoyuYtc7kd  come closer / ride me like a harley   

  album_release_date  album_total_tracks album_type  track_duration_min  
0         2025-10-31                   9      album                1.55  
1         2025-10-31                   1     single                3.07  
2         2025-10-31                   1     single                2.55  
3         2025-10-31                   9      album                1.69  
4         2025-10-30                   2     single                2.39

In [305]:
df2 = pd.read_csv("../data/raw/track_data_final.csv")

In [306]:
df2.head()

track_id                           track_name  track_number  \
0  6pymOcrCnMuCWdgGVTvUgP                                    3            57   
1  2lWc1iJlz2NVcStV5fbtPG                               Clouds             1   
2  1msEuwSBneBKpVCZQcFTsU  Forever & Always (Taylor’s Version)            11   
3  7bcy34fBT2ap1L4bfPsl9q            I Didn't Change My Number             2   
4  0GLfodYacy3BJE7AI3A8en                             Man Down             7   

   track_popularity  track_duration_ms  explicit     artist_name  \
0                61             213173     False  Britney Spears   
1                67             158760     False           BUNT.   
2                63             225328     False    Taylor Swift   
3                72             158463      True   Billie Eilish   
4                57             267013     False         Rihanna   

   artist_popularity  artist_followers      artist_genres  \
0               80.0        17755451.0            ['pop']   
1               69.0          293734.0  ['stutter house']   
2              100.0       145396321.0                 []   
3               90.0       118692183.0                 []   
4               90.0        68997177.0                 []   

                 album_id                   album_name album_release_date  \
0  325wcm5wMnlfjmKZ8PXIIn       The Singles Collection         2009-11-09   
1  2ArRQNLxf9t0O0gvmG5Vsj                       Clouds         2023-01-13   
2  4hDok0OAJd57SGIT8xuWJH  Fearless (Taylor's Version)         2021-04-09   
3  0JGOiO34nwfUdDrD612dOp            Happier Than Ever         2021-07-30   
4  5QG3tjE5L9F6O2vCAPph38                         Loud         2010-01-01   

   album_total_tracks   album_type  
0                  58  compilation  
1                   1       single  
2                  26        album  
3                  16        album  
4                  13        album

In [307]:
print("DF1 Columns:")
print(list(df1.columns))

print("\nDF2 Columns:")
print(list(df2.columns))

DF1 Columns:
['track_id', 'track_name', 'track_number', 'track_popularity', 'explicit', 'artist_name', 'artist_popularity', 'artist_followers', 'artist_genres', 'album_id', 'album_name', 'album_release_date', 'album_total_tracks', 'album_type', 'track_duration_min']

DF2 Columns:
['track_id', 'track_name', 'track_number', 'track_popularity', 'track_duration_ms', 'explicit', 'artist_name', 'artist_popularity', 'artist_followers', 'artist_genres', 'album_id', 'album_name', 'album_release_date', 'album_total_tracks', 'album_type']


In [308]:
print("DF1 shape:", df1.shape)
print("DF2 shape:", df2.shape)

DF1 shape: (8582, 15)
DF2 shape: (8778, 15)


- What This Does

Converts ms → minutes

Removes redundant column

Aligns column order with df1

In [309]:
# Convert milliseconds to minutes in df2
df2["track_duration_min"] = df2["track_duration_ms"] / 60000

# Drop old ms column
df2 = df2.drop(columns=["track_duration_ms"])

# Reorder columns to match df1
df2 = df2[df1.columns]

In [310]:
df2.head()

track_id                           track_name  track_number  \
0  6pymOcrCnMuCWdgGVTvUgP                                    3            57   
1  2lWc1iJlz2NVcStV5fbtPG                               Clouds             1   
2  1msEuwSBneBKpVCZQcFTsU  Forever & Always (Taylor’s Version)            11   
3  7bcy34fBT2ap1L4bfPsl9q            I Didn't Change My Number             2   
4  0GLfodYacy3BJE7AI3A8en                             Man Down             7   

   track_popularity  explicit     artist_name  artist_popularity  \
0                61     False  Britney Spears               80.0   
1                67     False           BUNT.               69.0   
2                63     False    Taylor Swift              100.0   
3                72      True   Billie Eilish               90.0   
4                57     False         Rihanna               90.0   

   artist_followers      artist_genres                album_id  \
0        17755451.0            ['pop']  325wcm5wMnlfjmKZ8PXIIn   
1          293734.0  ['stutter house']  2ArRQNLxf9t0O0gvmG5Vsj   
2       145396321.0                 []  4hDok0OAJd57SGIT8xuWJH   
3       118692183.0                 []  0JGOiO34nwfUdDrD612dOp   
4        68997177.0                 []  5QG3tjE5L9F6O2vCAPph38   

                    album_name album_release_date  album_total_tracks  \
0       The Singles Collection         2009-11-09                  58   
1                       Clouds         2023-01-13                   1   
2  Fearless (Taylor's Version)         2021-04-09                  26   
3            Happier Than Ever         2021-07-30                  16   
4                         Loud         2010-01-01                  13   

    album_type  track_duration_min  
0  compilation            3.552883  
1       single            2.646000  
2        album            3.755467  
3        album            2.641050  
4        album            4.450217

- Merge Both Data

In [311]:
df = pd.concat([df1, df2], ignore_index=True)

print("Merged shape:", df.shape)

Merged shape: (17360, 15)


- Remove Duplicates

In [312]:
df = df.drop_duplicates(subset=["track_id"])
print("After removing duplicates:", df.shape)

After removing duplicates: (8778, 15)


In [313]:
df

track_id  \
0      3EJS5LyekDim1Tf5rBFmZl   
1      1oQW6G2ZiwMuHqlPpP27DB   
2      7mdkjzoIYlf1rx9EtBpGmU   
3      67rW0Zl7oB3qEpD5YWWE5w   
4      15xptTfRBrjsppW0INUZjf   
...                       ...   
17117  06RKqktximm0kH8NVvxTyN   
17150  0ByMNEPAPpOR5H69DVrTNy   
17187  4pLwZjInHj3SimIyN9SnOz   
17227  6o3QUC5oAE4g6WxRIFcZtb   
17232  2E43WFS4rRc09za2r2GmZl   

                                              track_name  track_number  \
0                          Trippy Mane (ft. Project Pat)             4   
1                                                   OMG!             1   
2                                            Hard 2 Find             1   
3        Still Get Like That (ft. Project Pat & Starrah)             8   
4                                  ride me like a harley             2   
...                                                  ...           ...   
17117                                       Cherry Waves             4   
17150                               Don't Stop The Music             3   
17187                                       Side To Side             5   
17227                                       Wishing Well            14   
17232  My Songs Know What You Did In The Dark (Light ...             2   

       track_popularity  explicit    artist_name  artist_popularity  \
0                     0      True          Diplo               77.0   
1                     0      True       Yelawolf               64.0   
2                     4      True      Riff Raff               48.0   
3                    30      True          Diplo               77.0   
4                     0      True        Rumelis               48.0   
...                 ...       ...            ...                ...   
17117                64     False       Deftones               81.0   
17150                84     False        Rihanna               90.0   
17187                 2      True  Ariana Grande               90.0   
17227                 1      True     Juice WRLD               85.0   
17232                76     False   Fall Out Boy               77.0   

       artist_followers                                      artist_genres  \
0             2812821.0                                         moombahton   
1             2363438.0                  country hip hop, southern hip hop   
2              193302.0                                                NaN   
3             2813710.0                                         moombahton   
4                8682.0                                           dark r&b   
...                 ...                                                ...   
17117         7723018.0  ['nu metal', 'alternative metal', 'rap metal',...   
17150        69050055.0                                                 []   
17187       107418606.0                                            ['pop']   
17227        42565036.0                         ['melodic rap', 'emo rap']   
17232        11416476.0                                ['emo', 'pop punk']   

                     album_id                           album_name  \
0      5QRFnGnBeMGePBKF2xTz5z                   d00mscrvll, Vol. 1   
1      4SUmmwnv0xTjRcLdjczGg2                                 OMG!   
2      3E3zEAL8gUYWaLYB9L7gbp                          Hard 2 Find   
3      5QRFnGnBeMGePBKF2xTz5z                   d00mscrvll, Vol. 1   
4      06FDIpSHYmZAZoyuYtc7kd  come closer / ride me like a harley   
...                       ...                                  ...   
17117  4o1KnoVpzXZceJxyjELEQB                 Saturday Night Wrist   
17150  3JSWZWeTHF4HDGt5Eozdy7         Good Girl Gone Bad: Reloaded   
17187  4lVR2fg3DAUQpGVJ6DciHW                      Dangerous Woman   
17227  1R6vbGGXSEZZmTGn7ewwRL                    Legends Never Die   
17232  5jKMfS57mHTHzlSFGfPFxU                   Save Rock And Roll   

      album_release_date  album_total_tracks album_type  track_duration_min  
0             2025-10-31  

In [314]:
df.describe()

track_number  track_popularity  artist_popularity  artist_followers  \
count   8778.000000       8778.000000        8778.000000      8.778000e+03   
mean       5.749487         52.232855          69.934495      2.436033e+07   
std        6.017938         24.075221          19.573728      3.815353e+07   
min        1.000000          0.000000           0.000000      0.000000e+00   
25%        1.000000         39.000000          60.000000      5.128858e+05   
50%        4.000000         58.000000          74.000000      6.257170e+06   
75%        9.000000         71.000000          84.000000      3.055055e+07   
max      102.000000        100.000000         100.000000      1.455421e+08   

       album_total_tracks  track_duration_min  
count         8778.000000         8778.000000  
mean            13.779790            3.497875  
std             11.797943            1.055429  
min              1.000000            0.000000  
25%              6.000000            2.890000  
50%             13.000000            3.450000  
75%             17.000000            3.990000  
max            181.000000           13.510000

## Check Missing Values


In [315]:
df.isnull().sum()

track_id                 0
track_name               2
track_number             0
track_popularity         0
explicit                 0
artist_name              3
artist_popularity        0
artist_followers         0
artist_genres         3361
album_id                 0
album_name               2
album_release_date       0
album_total_tracks       0
album_type               0
track_duration_min       0
dtype: int64

artist_genres has 3361 missing values

Total dataset size ≈ 8778

That means:

3361
/
8778
≈
38
3361/8778≈38

So almost 40% of songs have no genre information.

We cannot simply drop those rows.

That would shrink dataset too much.

STEP 1 — Handle Missing Values Properly

In [316]:
# Replace missing text values
df["track_name"] = df["track_name"].fillna("Unknown Track")
df["artist_name"] = df["artist_name"].fillna("Unknown Artist")
df["album_name"] = df["album_name"].fillna("Unknown Album")

# Replace missing genres
df["artist_genres"] = df["artist_genres"].fillna("unknown")

# Verify
df.isnull().sum()

track_id              0
track_name            0
track_number          0
track_popularity      0
explicit              0
artist_name           0
artist_popularity     0
artist_followers      0
artist_genres         0
album_id              0
album_name            0
album_release_date    0
album_total_tracks    0
album_type            0
track_duration_min    0
dtype: int64

Inspect Raw Genre Values

In [317]:
df["artist_genres"].head(10)

0                                           moombahton
1                    country hip hop, southern hip hop
2                                              unknown
3                                           moombahton
4                                             dark r&b
5                                             dark r&b
6                                              unknown
7                                             dark r&b
8                                              unknown
9    nigerian drill, alté, afro adura, afrobeats, a...
Name: artist_genres, dtype: object

## Clean Genres

In [318]:
import ast

def clean_genres(genre):
    try:
        # If it's a list-like string, convert it
        if genre.startswith("["):
            genre_list = ast.literal_eval(genre)
            # Replace spaces with underscore in each genre
            genre_list = [g.replace(" ", "_") for g in genre_list]
            return " ".join(genre_list)
        else:
            # Single genre case
            return genre.replace(" ", "_")
    except:
        return "unknown"

df["clean_genres"] = df["artist_genres"].apply(clean_genres)

df[["artist_genres", "clean_genres"]].head()

artist_genres                       clean_genres
0                         moombahton                         moombahton
1  country hip hop, southern hip hop  country_hip_hop,_southern_hip_hop
2                            unknown                            unknown
3                         moombahton                         moombahton
4                           dark r&b                           dark_r&b

In [319]:
df["clean_genres"].head(20)

0                                            moombahton
1                     country_hip_hop,_southern_hip_hop
2                                               unknown
3                                            moombahton
4                                              dark_r&b
5                                              dark_r&b
6                                               unknown
7                                              dark_r&b
8                                               unknown
9     nigerian_drill,_alté,_afro_adura,_afrobeats,_a...
10                                              unknown
11                                                c-pop
12                                              unknown
13                                              unknown
14                                                c-pop
15                                              unknown
16                                              unknown
17                                              

## Select Numeric Features

In [320]:
numeric_features = [
    "track_popularity",
    "artist_popularity",
    "artist_followers",
    "track_duration_min"
]

df_numeric = df[numeric_features]
df_numeric.head()

track_popularity  artist_popularity  artist_followers  track_duration_min
0                 0               77.0         2812821.0                1.55
1                 0               64.0         2363438.0                3.07
2                 4               48.0          193302.0                2.55
3                30               77.0         2813710.0                1.69
4                 0               48.0            8682.0                2.39

## Scale Numeric Features

In [321]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
scaled_numeric = scaler.fit_transform(df_numeric)

scaled_numeric[:5]

array([[-2.16969265,  0.36098934, -0.56479008, -1.84568188],
       [-2.16969265, -0.30320404, -0.57656903, -0.40542697],
       [-2.00353725, -1.1206728 , -0.6334513 , -0.89814575],
       [-0.92352719,  0.36098934, -0.56476677, -1.71302682],
       [-2.16969265, -1.1206728 , -0.63829044, -1.04975153]])

## Vectorize Genres (TF-IDF)

In [322]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer(max_features=1000)

genre_matrix = tfidf.fit_transform(df["clean_genres"])

genre_matrix.shape

(8778, 610)

## Combine Numeric + Genre Features

In [323]:
from scipy.sparse import hstack

final_matrix = hstack([genre_matrix, scaled_numeric])

final_matrix.shape

(8778, 614)

Cosine Similarity

Why?

Because:

Works well with TF-IDF

Works well with high dimensional vectors

Scale independent

Standard for recommender systems

In [324]:
from sklearn.metrics.pairwise import cosine_similarity

similarity_matrix = cosine_similarity(final_matrix)

similarity_matrix.shape

(8778, 8778)

In [325]:
from sklearn.preprocessing import MinMaxScaler

pop_scaler = MinMaxScaler()

df["pop_score"] = pop_scaler.fit_transform(
    df[["track_popularity"]]
)

df[["track_name", "pop_score"]].head()

track_name  pop_score
0                    Trippy Mane (ft. Project Pat)       0.00
1                                             OMG!       0.00
2                                      Hard 2 Find       0.04
3  Still Get Like That (ft. Project Pat & Starrah)       0.30
4                            ride me like a harley       0.00

In [326]:
def assign_mood(row):
    genres = row["clean_genres"]
    pop = row["track_popularity"]
    
    if any(g in genres for g in ["rap", "hip_hop", "synthpop", "pop"]) and pop > 60:
        return "Energetic"
    
    elif any(g in genres for g in ["indie", "soft_pop", "folk", "r&b"]):
        return "Chill"
    
    elif any(g in genres for g in ["metal", "dark", "soundtrack", "alternative"]):
        return "Emotional"
    
    elif pop > 80:
        return "Trending"
    
    else:
        return "Neutral"

df["mood"] = df.apply(assign_mood, axis=1)

df["mood"].value_counts()

mood
Neutral      5300
Energetic    1749
Emotional     748
Chill         712
Trending      269
Name: count, dtype: int64

## Create Recommendation Function

In [327]:
import numpy as np
from sklearn.metrics.pairwise import cosine_similarity

def recommend(song_name, top_n=5, alpha=0.8, mood=None):
    
    idx = df[df["track_name"] == song_name].index[0]
    
    # Compute similarity
    sim_scores = cosine_similarity(
        final_matrix[idx], 
        final_matrix
    ).flatten()
    
    # Blend similarity + popularity
    blended_scores = alpha * sim_scores + (1 - alpha) * df["pop_score"]
    
    # Create temp dataframe
    temp_df = df.copy()
    temp_df["score"] = blended_scores
    
    # Filter by mood if provided
    if mood:
        temp_df = temp_df[temp_df["mood"] == mood]
    
    # Sort by score
    temp_df = temp_df.sort_values("score", ascending=False)
    
    # Remove itself
    temp_df = temp_df[temp_df["track_name"] != song_name]
    
    return temp_df[
        ["track_name", "artist_name", "mood", "track_popularity"]
    ].head(top_n)

In [328]:
print(df.columns)
df["mood"] = df.apply(assign_mood, axis=1)
print("mood" in df.columns)

Index(['track_id', 'track_name', 'track_number', 'track_popularity',
       'explicit', 'artist_name', 'artist_popularity', 'artist_followers',
       'artist_genres', 'album_id', 'album_name', 'album_release_date',
       'album_total_tracks', 'album_type', 'track_duration_min',
       'clean_genres', 'pop_score', 'mood'],
      dtype='object')
True


In [329]:
final_matrix = final_matrix.tocsr()

In [330]:
recommend("OMG!")

track_name           artist_name     mood  \
8904                              SNAP             Rosa Linn  Neutral   
14856  Waves - Robin Schulz Radio Edit             Mr. Probz  Neutral   
14089                          Hayloft         Mother Mother  Neutral   
9030                  Sex, Drugs, Etc.         Beach Weather  Neutral   
9211                   Fingers Crossed  Lauren Spencer Smith  Neutral   

       track_popularity  
8904                  9  
14856                 8  
14089                 3  
9030                  0  
9211                  0

In [331]:
recommend("Cherry Waves")

track_name artist_name  \
6930                                       Beauty School    Deftones   
5643                                      The Devil in I    Slipknot   
12496                                              Messy  Lola Young   
8047             Bitter Sweet Symphony - Remastered 2016   The Verve   
14838  I'm the One (feat. Justin Bieber, Quavo, Chanc...   DJ Khaled   

            mood  track_popularity  
6930   Energetic                67  
5643   Energetic                73  
12496   Trending                83  
8047   Energetic                83  
14838    Neutral                71

In [332]:
recommend("Cherry Waves", alpha=0.8)

track_name artist_name  \
6930                                       Beauty School    Deftones   
5643                                      The Devil in I    Slipknot   
12496                                              Messy  Lola Young   
8047             Bitter Sweet Symphony - Remastered 2016   The Verve   
14838  I'm the One (feat. Justin Bieber, Quavo, Chanc...   DJ Khaled   

            mood  track_popularity  
6930   Energetic                67  
5643   Energetic                73  
12496   Trending                83  
8047   Energetic                83  
14838    Neutral                71

In [333]:
recommend("Cherry Waves", alpha=0.5)

track_name        artist_name       mood  \
12496                                    Messy         Lola Young   Trending   
8047   Bitter Sweet Symphony - Remastered 2016          The Verve  Energetic   
8185                             Fade Into You         Mazzy Star  Energetic   
8030                                      Iris  The Goo Goo Dolls   Trending   
2034                                 About You           The 1975   Trending   

       track_popularity  
12496                83  
8047                 83  
8185                 86  
8030                 92  
2034                 86

In [334]:
df["clean_genres"].value_counts().head(10)

clean_genres
unknown                                        3361
pop                                             351
soundtrack                                      346
country,_pop,_indie,_folk                       324
soft_pop                                        209
rap                                             192
r&b_pop,_alternative_r&b,_hip_hop,_synthpop     141
                                                111
dark_r&b                                        111
alternative_pop,_indie_pop                       99
Name: count, dtype: int64

In [335]:
import pickle
import os

os.makedirs("../artifacts", exist_ok=True)

pickle.dump(df, open("../artifacts/df.pkl", "wb"))
pickle.dump(final_matrix, open("../artifacts/final_matrix.pkl", "wb"))
pickle.dump(tfidf, open("../artifacts/tfidf.pkl", "wb"))
pickle.dump(scaler, open("../artifacts/scaler.pkl", "wb"))